bundles/org.openhab.core.config.discovery.usbserial.ser2net/src/main/java/org/openhab/core/config/discovery/usbserial/ser2net/internal/Ser2NetUsbSerialDiscovery.java

```diff
@@ -0,0 +1,206 @@
+
+import javax.jmdns.ServiceEvent;
+import javax.jmdns.ServiceInfo;
+import javax.jmdns.ServiceListener;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.core.config.discovery.usbserial.UsbSerialDeviceInformation;
+import org.openhab.core.config.discovery.usbserial.UsbSerialDiscovery;
+import org.openhab.core.config.discovery.usbserial.UsbSerialDiscoveryListener;
+import org.openhab.core.io.transport.mdns.MDNSClient;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link UsbSerialDiscovery} that implements background discovery of RFC2217 by listening to
+ * ser2net mDNS service events.
+ *
+ * @author Wouter Born - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = UsbSerialDiscovery.class)
+public class Ser2NetUsbSerialDiscovery implements ServiceListener, UsbSerialDiscovery {
+
+    private final Logger logger = LoggerFactory.getLogger(Ser2NetUsbSerialDiscovery.class);
+
+    static final String SERVICE_TYPE = "_iostream._tcp.local.";
+
+    static final String PROPERTY_PROVIDER = "provider";
+    static final String PROPERTY_DEVICE_TYPE = "devicetype";
+    static final String PROPERTY_GENSIO_STACK = "gensiostack";
+
+    static final String PROPERTY_VENDOR_ID = "idVendor";
+    static final String PROPERTY_PRODUCT_ID = "idProduct";
+
+    static final String PROPERTY_SERIAL_NUMBER = "serial";
+    static final String PROPERTY_MANUFACTURER = "manufacturer";
+    static final String PROPERTY_PRODUCT = "product";
+
+    static final String PROPERTY_INTERFACE_NUMBER = "bInterfaceNumber";
+    static final String PROPERTY_INTERFACE = "interface";
+
+    static final String SER2NET = "ser2net";
+    static final String SERIALUSB = "serialusb";
+    static final String TELNET_RFC2217_TCP = "telnet(rfc2217),tcp";
+
+    static final Duration SINGLE_SCAN_DURATION = Duration.ofSeconds(5);
+    static final String SERIAL_PORT_NAME_FORMAT = "rfc2217://%s:%s";
+
+    private final Set<UsbSerialDiscoveryListener> discoveryListeners = new CopyOnWriteArraySet<>();
+    private final MDNSClient mdnsClient;
+
+    private boolean backgroundDiscoveryEnabled = false;
+
+    private Set<UsbSerialDeviceInformation> lastScanResult = new HashSet<>();
+
+    @Activate
+    public Ser2NetUsbSerialDiscovery(final @Reference MDNSClient mdnsClient) {
+        this.mdnsClient = mdnsClient;
+    }
+
+    @Override
+    public void registerDiscoveryListener(UsbSerialDiscoveryListener listener) {
+        discoveryListeners.add(listener);
+    }
+
+    @Override
+    public void unregisterDiscoveryListener(UsbSerialDiscoveryListener listener) {
+        discoveryListeners.remove(listener);
+    }
+
+    @Override
+    public synchronized void startBackgroundScanning() {
+        backgroundDiscoveryEnabled = true;
```

Comment:
Setting this flag here has an irritating effect on my understanding. Isn't it sufficient to unregister it from listening to the mdns client?

Reply:
What is exactly the irritating effect?

It keeps track of this so that any results the `mdnsClient` produces are guaranteed to be ignored whenever it is instructed to stop scanning.

Reply:
Maybe it is just the naming. The prefix `Enabled`. Rather use `backgroundDiscoberyRunning`?

Reply:
I used this name because its behaviour is very similar to the background scanning in the `AbstractDiscoveryService` :slightly_smiling_face:  : 

https://github.com/openhab/openhab-core/blob/5d5b7665b4b64f1e50e4e841557847d746915fb9/bundles/org.openhab.core.config.discovery/src/main/java/org/openhab/core/config/discovery/AbstractDiscoveryService.java#L64

Reply:
I think there is a little misunderstanding from your side. `backgroundDiscoveryEnabled` in `AbstractDiscoveryService` allowed a user to enable/disabled background discovery actions. It does not keep track if a discovery job is running or not - which is the goal you want to archive. Isn't it?

Reply:
How about calling it something like `sendAnnouncements` or `notifyListeners` That's because the discovery can still be running when you set the boolean and its purpose it to keep track of this.

Reply:
Yes, go for it. I'd prefer `notifyListeners`.

Reply:
OK done. :slightly_smiling_face: